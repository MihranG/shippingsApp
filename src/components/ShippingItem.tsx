import * as React from 'react'
import { Field, reduxForm } from 'redux-form'
import {TextField, Button, Paper, Typography, makeStyles, Theme, createStyles} from '@material-ui/core'
import { connect, compose } from 'react-redux'
import { IState } from '../interfaces'
import {RouteComponentProps} from 'react-router-dom'
import { editShipmentWithId } from '../store'


interface IComponentProps extends RouteComponentProps<{}> {

}

interface IStateProps {
    itemValues: any
    formValues: any
    editShippingItem: (id: number, shipmentDetails: any)=>void
}

interface IProps extends IComponentProps, IStateProps{
    
}


export const useStyles = makeStyles((theme: Theme) =>
  createStyles({
  
    paper: {
      marginBottom: theme.spacing(2),
      padding: 15
    },
    form: {
      width: 500
    },
    button: {
      margin: 15
    },
    type: {
      margin: '15px 0'
    },
    visuallyHidden: {
      border: 0,
      clip: 'rect(0 0 0 0)',
      height: 1,
      margin: -1,
      overflow: 'hidden',
      padding: 0,
      position: 'absolute',
      top: 20,
      width: 1,
    },
  }),
);

const validate = values => {
    const errors = {}
    const requiredFields = [
      'name',
      'destination',
      'mode',
      'origin',
      'type'
    ]
    requiredFields.forEach(field => {
      if (!values[field]) {
        errors[field] = 'Required'
      }
    })
    
    return errors
  }
// @ts-ignore
export const EditShipmentFormDisconnected: React.FC<IProps> = ({editShippingItem, match, formValues, itemValues , history, ...rest})=>{
    const handleFormSubmit = (e:React.FormEvent<HTMLFormElement>)=>{
        e.preventDefault();
        editShippingItem(match.params.id, formValues.values);
        history.replace('/')
    }

    const handleFormCancel = ()=>{
      history.replace('/')
    }

    const classes = useStyles();
    return (
        <form onSubmit={handleFormSubmit} className={classes.form}>
          <Paper elevation={3} className={classes.paper} >
            <Typography component='h3'> 
                  ID: {itemValues.id}
            </Typography>
            <div>
                <Field
                    name="name"
                    component={renderTextField}
                    label="First Name"
                />
            </div>
            <div>
                <Field
                    name="destination"
                    component={renderTextField}
                    label="Destination"
                />
            </div>
            <div>
                <Field
                    name="origin"
                    component={renderTextField}
                    label="Origin"
                />
            </div>
            <Typography component='h3' className={classes.type}> 
                Type: {itemValues.type}
            </Typography>
            <Button disabled={(formValues && (!formValues.anyTouched || !!formValues.syncErrors ))} type='submit' variant='contained' className={classes.button}> Change</Button>
            <Button type='reset' variant='text' onClick={handleFormCancel}> Cancel</Button>
          </Paper>
        </form>
    )
}


const renderTextField = ({
    label,
    input,
    meta: { touched, invalid, error },
    ...custom
  }) => (
    <TextField
      label={label}
      placeholder={label}
      error={touched && invalid}
      helperText={touched && error}
      {...input}
      {...custom}
    />
  )


const mapStateToProps :(state: IState, ownProps : RouteComponentProps)=>{itemValues: object, formValues: object} = (state, ownProps)=>{
  const itemValues = state.shipments.data[ownProps.match.params.id];
  return({
      itemValues,
      formValues: state.form.editShipping,
      initialValues: itemValues
    })
  }

const mapDispatchToProps =  (dispatch) =>({
    editShippingItem: (id, shipmentDetails)=>dispatch(editShipmentWithId({id, shipmentDetails}))
})



export default connect(mapStateToProps, mapDispatchToProps)(
  reduxForm({
    form: 'editShipping',
    validate
  })(EditShipmentFormDisconnected)
)
