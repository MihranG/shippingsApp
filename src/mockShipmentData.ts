import {IShipment} from './interfaces'

export const data: IShipment[]  =
    [{
        "id": "S1000",
        "name": "TEST1",
        "cargo": [
          {
            "type": "Fabric",
            "description": "1000 Blue T-shirts",
            "volume": "2"
          },
          {
            "type": "Fabric",
            "description": "2000 Green T-shirts",
            "volume": "3"
          }
        ],
        "mode": "sea",
        "type": "FCL",
        "destination": "Saarbrücker Str. 38, 10405 Berlin",
        "origin": "Shanghai Port",
        "services": [
          {
            "type": "customs"
          }
        ],
        "total": "1000",
        "status": "ACTIVE",
        "userId": "U1000"
      },
      {
        "id": "S1001",
        "name": "TEST2",
        "cargo": [
          {
            "type": "Furniture",
            "description": "300 Tables",
            "volume": "20"
          },
          {
            "type": "Furniture",
            "description": "1500 Chairs",
            "volume": "15"
          }
        ],
        "mode": "sea",
        "type": "FCL",
        "destination": "Saarbrücker Str. 38, 10405 Berlin",
        "origin": "Ningbo port",
        "services": [
          {
            "type": "customs"
          },
          {
            "type": "insurance",
            "value": "100"
          }
        ],
        "total": "3000",
        "status": "ACTIVE",
        "userId": "U1002"
      }
  ]

export default data