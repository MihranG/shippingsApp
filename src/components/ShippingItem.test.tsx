import * as React from 'react';
import { shallow, mount } from 'enzyme';
import shippingItem from './ShippingItem'
import ShippingItem from './ShippingItem';
import {data as mockData}  from '../mockShipmentData'
import Provider from 'react-redux/lib/components/Provider';
import configureStore from "redux-mock-store";



const normalizedData = {};
mockData.forEach(item=>{
  normalizedData[item.id] = item
})



const mockStore = configureStore();
const store = mockStore({
    shipments: {data: normalizedData}
})



describe('shippingItem Component',()=>{
    let wrapper;
    const matchPropMock = {
        params: {
            id: 'S1000'
        }
    }
    // console.log(133,store.getState())
    beforeAll(()=>{
        wrapper = shallow(
            <Provider store={store}>
                <ShippingItem itemValues={mockData[0]} match={matchPropMock} />
            </Provider>
        )
    })
    test('should match snapshot',()=>{
        expect(wrapper).toMatchSnapshot()
    })
    afterAll(()=>{
        wrapper.detach()
    })
})