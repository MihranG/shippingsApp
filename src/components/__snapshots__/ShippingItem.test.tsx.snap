// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shippingItem Component should match snapshot 1`] = `ShallowWrapper {}`;
